File: Solutions/Problem2/Solution2.java
import java.util.Arrays;

public class Solution2 {
    // User input
    public static int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (nums[mid] == target) {
                return mid;
            }
            
            if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return -1;
    }
    // End of user input

    static class Checker {
        public static void check(int[] nums, int target, int expected) {
            int result = search(nums, target);
            if (result == expected) {
                System.out.printf("Test Case Passed: Result = %d, Expected = %d%n", 
                                result, expected);
            } else {
                System.out.printf("Failed Test Case: Result = %d, Expected = %d%n", 
                                result, expected);
            }
        }
    }

    private static int[] parseArray(String str) {
        str = str.substring(1, str.length() - 1); // Remove [ ]
        String[] parts = str.split(",");
        int[] nums = new int[parts.length];
        
        for (int i = 0; i < parts.length; i++) {
            nums[i] = Integer.parseInt(parts[i].trim());
        }
        return nums;
    }

    public static void main(String[] args) {
        if (args.length != 3) {
            System.err.printf("Usage: java Solution2 <array> <target> <expected>%n");
            System.exit(1);
        }

        int[] nums = parseArray(args[0]);
        int target = Integer.parseInt(args[1]);
        int expected = Integer.parseInt(args[2]);

        Checker.check(nums, target, expected);
    }
} 

File: Solutions/Problem2/solution2.rs
use std::env;

// User input
fn search(nums: &[i32], target: i32) -> i32 {
    let mut left = 0;
    let mut right = nums.len() as i32 - 1;
    
    while left <= right {
        let mid = left + (right - left) / 2;
        let mid_idx = mid as usize;
        
        if nums[mid_idx] == target {
            return mid;
        }
        
        if nums[mid_idx] < target {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    -1
}
// End of user input

struct Checker;

impl Checker {
    fn check(func: fn(&[i32], i32) -> i32, nums: &[i32], target: i32, expected: i32) {
        let result = func(nums, target);
        if result == expected {
            println!("Test Case Passed: Result = {}, Expected = {}", result, expected);
        } else {
            println!("Failed Test Case: Result = {}, Expected = {}", result, expected);
        }
    }
}

fn parse_array(input: &str) -> Vec<i32> {
    input[1..input.len()-1]  // Remove [ ]
        .split(',')
        .map(|s| s.parse().unwrap())
        .collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 4 {
        eprintln!("Usage: {} <array> <target> <expected>", args[0]);
        std::process::exit(1);
    }

    let nums = parse_array(&args[1]);
    let target = args[2].parse().expect("Invalid target");
    let expected = args[3].parse().expect("Invalid expected value");

    Checker::check(search, &nums, target, expected);
} 